### PR TITLE
[2.6.0 stabilization] add SurfaceMeshesToPlanes, based on #7555

### PR DESCRIPTION
**Pull request by davidkline-ms**
This change tales SurfaceMeshesToPlanes.cs from #7555 and adds it to SDK/Experimental.

Adds protection for missing native DLL. Uses the to be published MRTK PlaneFinding package.

### CODE REVIEW

**Comment by davidkline-ms** [comment]
Opened as DRAFT to get feedback while testing.

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:18]: Is there an existing `BaseSpatialAwarenessObject` we could use for this? There's `SpatialAwarenessPlanarObject`, but it doesn't have a type. Should we extend it to `SpatialAwarenessPlanarObjectWithType`? Or just add a type to the existing class, which may or may not be set

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:273]: How does this script aquire these meshes?

Comment (keveleigh) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:273]: Should we make it a mesh event listener?

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:50]: This all feels a little observer-y 😬 do we think this is the right path, or should we go ahead and add a data provider observer instead?

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:174]: This all feels a little observer-y 😬 do we think this is the right path, or should we go ahead and add a data provider observer instead?

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (davidkline-ms) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:273]: I am going to leave that up to the developer. In Holograms 230, we collected meshes for n seconds and then called the MakePlanes method. For a demo, I am considering adding a voice command to generate meshes.

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (davidkline-ms) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:174]: Not sure about that. @srinjoym had an observer PR as well.

I do like the thought of this being standalone, but also willing to reconsider and push this change to post 2.6.0

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (keveleigh) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:273]: Ahh I see. I missed the part where this script was specifically grabbing the `SpatialAwarenessObjectParent`.

Comment (keveleigh) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:273]: >I am considering adding a voice command to generate meshes.

Yeah, a voice command or even just a button seems reasonable

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (davidkline-ms) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:18]: i will double check. since this is experimental, should it block this PR?

Comment (keveleigh) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:18]: I guess since this struct is private, we can leave it as-is. I'm going to push some changes to make the struct a little more strict

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:192]: This is inside an `#if PLANE_FINDING_PRESENT` on line 169. I'll move it out in my upcoming commit

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (davidkline-ms) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs]: fixed

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:196]: merge markers here. you can remove 

```c#
private float floorYPosition;
private float ceilingYPosition;
```

since they're now down at line 200

Comment (davidkline-ms) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:196]: thx. good catch!

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:333]: Should this section use `SpatialAwarenessPlanarObject.CreateSpatialObject` instead of the explicit constructor? (Also almost feels like a weird pattern to still have a public constructor with the `CreateSpatialObject` method, but that's not relevant here)

Comment (keveleigh) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:333]: There's also some overlap between `ConfigurePlaneGameObject` and `SpatialAwarenessPlanarObject.CreateSpatialObject` that'd we'd be able to dedupe and reuse existing code!

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (keveleigh) [inline Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:4]: Is this used?

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (keveleigh) [inline Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:25]: Does this need to be publicly settable? Feels like info the observer would provide when the object is created, as opposed to needing to be settable elsewhere

Comment (davidkline-ms) [reply Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:25]: you are probably right :)

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (davidkline-ms) [reply Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:4]: i'll double check and remove

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (davidkline-ms) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:333]: good call

File: Documentation/SpatialAwareness/SurfaceMeshesToPlanes.md
Comment (MaxWang-MS) [inline Documentation/SpatialAwareness/SurfaceMeshesToPlanes.md:3]:
```suggestion
`SurfaceMeshesToPlanes.cs` is a new script that enables simple plane finding applications in MRTK. This script processes mesh data from the Spatial Awareness system and builds planes for Floors, Ceilings, Walls and Platforms in a scene. This script is fully compatible with HoloLens 1 and 2.
```

File: Documentation/SpatialAwareness/SurfaceMeshesToPlanes.md
Comment (MaxWang-MS) [inline Documentation/SpatialAwareness/SurfaceMeshesToPlanes.md:17]: Would it be good if we mention the sample scene somewhere in this file?

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (davidkline-ms) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs]: updated

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (davidkline-ms) [reply Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:4]: removed.

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (davidkline-ms) [reply Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs]: fixed

File: Documentation/SpatialAwareness/SurfaceMeshesToPlanes.md
Comment (davidkline-ms) [reply Documentation/SpatialAwareness/SurfaceMeshesToPlanes.md:23]: added

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (keveleigh) [reply Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:38]: `SurfaceMeshesToPlanes` should still be able to edit the provided GameObject to its liking after

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (davidkline-ms) [reply Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:23]: good catch. forgot to mark that private

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (davidkline-ms) [reply Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:38]: i'll take another spin :smile:

File: Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:329]: Can extents param of `CreateSpatialObject` be changed back to the original parameter of `size` and we pass `PlaneThickness` as the z value? that way we don't have to manually set it below on line 335?

Comment (keveleigh) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:329]: Though, i guess this was is an "either way" thing. I could see us only wanting extents and providing a standard thickness in the implementation

Comment (davidkline-ms) [reply Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs:329]: sure!

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (keveleigh) [inline Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:62]: This feels potentially implementation-specific

**Review by keveleigh — APPROVED** [review]
Haven't tried running on-device, but assuming that works, I think we can go ahead and merge this as experimental!

File: Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs
Comment (davidkline-ms) [reply Assets/MRTK/Core/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs:62]: by default, shadows off are a perf benefit.

**Pull request merged**
